core/trino-main/src/main/java/io/trino/operator/TopNRankingOperator.java

```diff
@@ -291,35 +362,53 @@ public Page getOutput()
             unfinishedWork = null;
         }
 
-        if (!finishing) {
+        if (!finishing && (!partial || groupedTopNBuilder == null || !groupedTopNBuilder.isFull()) && outputIterator == null) {
```

Comment:
If you add the isBuilderFull() method I suggested above, this would just be:

```
if (!finishing && !(partial && isBuilderFull()) && outputIterator == null) {
```
which I think reads much better

Reply:
Done